src/hotspot/cpu/x86/peephole_x86_64.cpp

```diff
@@ -235,6 +236,132 @@ bool Peephole::test_may_remove(Block* block, int block_index, PhaseCFG* cfg_, Ph
+// if all leaP*s and the decode share the same parent.
+// Additionally, if the register allocator spills the result of the LoadN we can get such a graph:
+//               LoadN
+//                 |
+//        DefinitionSpillCopy
+//           /           \
+// MemToRegSpillCopy   MemToRegSpillCopy
+//           |           /
+//           | decodeHeapOop_not_null
+//           |   /              \
+//           leaP*          MachProj (leaf)
+// In this case where te common parent of the leaP* and the decode is one MemToRegSpill Copy
+// away, this peephole can als recognize the decode as redundant and also remove the spill copy
+// if that is only used by the decode.
+bool Peephole::lea_remove_redundant(Block* block, int block_index, PhaseCFG* cfg_, PhaseRegAlloc* ra_,
+                                    MachNode* (*new_root)(), uint inst0_rule) {
+  MachNode* lea_derived_oop = block->get_node(block_index)->as_Mach();
+  assert(lea_derived_oop->rule() == inst0_rule, "sanity");
+  assert(lea_derived_oop->ideal_Opcode() == Op_AddP, "sanity");
+
+  MachNode* decode = lea_derived_oop->in(AddPNode::Base)->isa_Mach();
+  if (decode == nullptr || decode->ideal_Opcode() != Op_DecodeN) {
+    return false;
+  }
+
+  // Check that the lea and the decode live in the same block.
+  if (!block->contains(decode)) {
+    return false;
+  }
+
+  bool is_spill = lea_derived_oop->in(AddPNode::Address) != decode->in(1) &&
+                  lea_derived_oop->in(AddPNode::Address)->is_SpillCopy() &&
+                  decode->in(1)->is_SpillCopy();
+
+  // The leaP* and the decode must have the same parent. If we have a spill, they must have
+  // the same grandparent.
+  if ((!is_spill && lea_derived_oop->in(AddPNode::Address) != decode->in(1)) ||
+      (is_spill && lea_derived_oop->in(AddPNode::Address)->in(1) != decode->in(1)->in(1))) {
+    return false;
+  }
+
+  // Ensure the decode only has the leaP*s with the same (grand)parent and a MachProj leaf as children.
+  MachProjNode* proj = nullptr;
+  for (DUIterator_Fast imax, i = decode->fast_outs(imax); i < imax; i++) {
+    Node* out = decode->fast_out(i);
+    if (out == lea_derived_oop) {
+      continue;
+    }
+    if (out->is_MachProj() && out->outcnt() == 0) {
+      proj = out->as_MachProj();
+      continue;
+    }
+    if (out->is_Mach()) {
+      MachNode* other_lea = out->as_Mach();
+      if ((other_lea->rule() == leaP32Narrow_rule ||
+           other_lea->rule() == leaP8Narrow_rule ||
+           other_lea->rule() == leaPCompressedOopOffset_rule) &&
+           other_lea->in(AddPNode::Base) == decode &&
+          (other_lea->in(AddPNode::Address) == decode->in(1) ||
+          (is_spill && other_lea->in(AddPNode::Address)->in(1) == decode->in(1)->in(1)))) {
+        continue;
+      }
+    }
+    // There is other stuff we do not expect...
+    return false;
+  }
+
+  // Ensure the MachProj is in the same block as the decode and the lea.
+  if (!block->contains(proj)) {
```

Comment:
Should we check `proj == nullptr` ?

Reply:
Indeed, we should. Thank you for pointing it out.

I fixed it in [bf75c0d](https://github.com/openjdk/jdk/pull/25471/commits/bf75c0da751def40149b5548fd0c89595318fc11).